### PR TITLE
Exposing stream_format params to user

### PR DESCRIPTION
**Pull request by Arun-Prasad-V**
The user can select the _stream_format_ for all the video streams through below params:
- rgb_camera.color_format
- depth_module.depth_format
- depth_module.infra_format
- depth_module.infra1_format
- depth_module.infra2_format

Run `ros2 param describe <your_node_name> <param_name>` command to know the list of formats supported by that particular stream. 
Alternatively, run `rs-enumerate-devices` command to know the list of profiles supported by the sensors connected.

### CODE REVIEW

File: realsense2_camera/src/profile_manager.cpp [hunk 1]
```diff
@@ -160,6 +160,21 @@ void ProfilesManager::addWantedProfiles(std::vector<rs2::stream_profile>& wanted
             ROS_DEBUG_STREAM("Found profile for " << ros_stream_to_string(sip.first) << ":" << sip.second);
         }
     }
+
+    // Warn the user if the enabled stream cannot be opened due to wrong profile selection
+    for (auto const & x : _enabled_profiles)
```
Comment (Arun-Prasad-V) [inline realsense2_camera/src/profile_manager.cpp]: During enabling / re-enabling the stream, warn the user if the enabled stream cannot be opened due to wrong profile selection.

File: realsense2_camera/launch/rs_multi_camera_launch.py [hunk 2]
```diff
@@ -33,8 +33,8 @@
 sys.path.append(str(pathlib.Path(__file__).parent.absolute()))
 import rs_launch
 
-local_parameters = [{'name': 'camera_name1', 'default': 'camera1', 'description': 'camera unique name'},
-                    {'name': 'camera_name2', 'default': 'camera2', 'description': 'camera unique name'},
+local_parameters = [{'name': 'camera_name_1', 'default': 'camera1', 'description': 'camera unique name'},
```
Comment (SamerKhshiboun) [inline realsense2_camera/launch/rs_multi_camera_launch.py]: is this change required for this feature ?

File: realsense2_camera/src/profile_manager.cpp [hunk 3]
```diff
@@ -160,6 +160,21 @@ void ProfilesManager::addWantedProfiles(std::vector<rs2::stream_profile>& wanted
...
+    {
+        auto sip = x.first;
+        auto stream_enabled = x.second;
+
+        if (*stream_enabled && !found_sips[sip])
+        {
+            ROS_WARN_STREAM("Couldn't open " << ros_stream_to_string(sip.first) << ":" << sip.second << " stream "
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/profile_manager.cpp:173]: Please add this information to the readme, maybe under a new section/part about profile formats

File: realsense2_camera/src/profile_manager.cpp [hunk 4]
```diff
@@ -275,7 +288,48 @@ std::string VideoProfilesManager::get_profiles_descriptions()
...
+        descriptors_strm << profile_format_str << "\n";
+    }
+    std::string descriptors(descriptors_strm.str());
+    descriptors.pop_back();
+    return descriptors;
+}
+
+void VideoProfilesManager::registerVideoSensorProfileFormat(stream_index_pair sip)
+{
+    if (sip == DEPTH)
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/profile_manager.cpp:320]: should these formats be hard coded? what will happen if different camera models support different default profiles?
can't we use liberalsense api to fetch/query the default profiles for each sip ?

File: realsense2_camera/launch/rs_launch.py [hunk 5]
```diff
@@ -31,8 +31,13 @@
                            {'name': 'log_level',                    'default': 'info', 'description': 'debug log level [DEBUG|INFO|WARN|ERROR|FATAL]'},
                            {'name': 'output',                       'default': 'screen', 'description': 'pipe node output [screen|log]'},
                            {'name': 'depth_module.profile',         'default': '0,0,0', 'description': 'depth module profile'},                           
+                           {'name': 'depth_module.profile.depth_stream_format',         'default': 'Z16', 'description': 'depth stream format'},
```
Comment (SamerKhshiboun) [inline realsense2_camera/launch/rs_launch.py]: please explain about these parameter in the Readme

File: realsense2_camera/src/profile_manager.cpp (hunk 1 above)
Comment (SamerKhshiboun) [reply realsense2_camera/src/profile_manager.cpp]: would prefer to rename "x" to something more understandable (like "profile")

File: realsense2_camera/src/profile_manager.cpp [hunk 6]
```diff
@@ -275,7 +288,48 @@ std::string VideoProfilesManager::get_profiles_descriptions()
     return descriptors;
 }
 
-void VideoProfilesManager::registerVideoSensorParams()
+std::string VideoProfilesManager::get_profile_formats_descriptions(stream_index_pair sip)
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/profile_manager.cpp]: I would keep "camel case" convention of function members for this class, and in general in the whole file.

I know there is a function get_profile_descriptions() and profile_string() that arenot a camel case aligned, so maybe we need to rename them too :) (and you can ignore them for now)

File: realsense2_camera/launch/rs_multi_camera_launch.py (hunk 2 above)
Comment (Arun-Prasad-V) [reply realsense2_camera/launch/rs_multi_camera_launch.py]: No, its not related.

Here, '1' and '2' are appended as suffix of actual param names. For `enable_infra` param, when '1' is appended, it becomes `enable_infra1` and we already have an original param with that name. That leads to misconfiguration of params.

File: realsense2_camera/src/profile_manager.cpp (hunk 4 above)
Comment (Arun-Prasad-V) [reply realsense2_camera/src/profile_manager.cpp:320]: Sure. Will do so.

File: realsense2_camera/launch/rs_multi_camera_launch.py (hunk 2 above)
Comment (SamerKhshiboun) [reply realsense2_camera/launch/rs_multi_camera_launch.py]: I am afraid that many users use the multi camera launch file, and maybe they rely on that parameters.
let me check that tomorrow

File: realsense2_camera/src/profile_manager.cpp (hunk 4 above)
Comment (Arun-Prasad-V) [reply realsense2_camera/src/profile_manager.cpp:320]: LibRealSense APIs gives the default profiles of color and depth streams only. Couldn't get those for Infra streams.

File: realsense2_camera/launch/rs_multi_camera_launch.py (hunk 2 above)
Comment (SamerKhshiboun) [reply realsense2_camera/launch/rs_multi_camera_launch.py]: Hi @Arun-Prasad-V ,
Ok.. I see there is no other way to do after adding the enable_infra capability.
Can you please update the readme section with these new parameters ? specially for multi camera launching ?

File: realsense2_camera/src/profile_manager.cpp (hunk 4 above)
Comment (SamerKhshiboun) [reply realsense2_camera/src/profile_manager.cpp:320]: @Arun-Prasad-V,
can we query the device about default profiles for the given streams, and if default is not given, then to hard coded it ?

File: realsense2_camera/launch/rs_multi_camera_launch.py (hunk 2 above)
Comment (Arun-Prasad-V) [reply realsense2_camera/launch/rs_multi_camera_launch.py]: @SamerKhshiboun ,
I have moved this part alone separate PR https://github.com/IntelRealSense/realsense-ros/pull/2822.
Will track it there.

File: realsense2_camera/src/profile_manager.cpp (hunk 4 above)
Comment (Arun-Prasad-V) [reply realsense2_camera/src/profile_manager.cpp:320]: Yes, we can query about default profiles and configure the format, if found. If not, maybe can we choose the first found format whose profile's resolution matches with depth stream? So that there won't be anything hard coded.

File: realsense2_camera/src/profile_manager.cpp [hunk 7]
```diff
@@ -333,12 +406,69 @@ void VideoProfilesManager::registerVideoSensorParams()
...
+                                                    << "Setting ROS param back to: " << crnt_profile_str.str());
                     }
                     _params.getParameters()->queueSetRosValue(parameter.get_name(), crnt_profile_str.str());
                 }
             }, crnt_descriptor);
     _parameters_names.push_back(param_name);
+
+    for (auto sip : sips)
+    {
+        std::string param_name(_module_name + ".profile." + STREAM_NAME(sip) + "_stream_format");
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/profile_manager.cpp]: can we work with a shorter name ?
like
depth_module.infra1_format ?
I found that long names are not very easy to understand, and can easily used in a wrong way.

**Comment by SamerKhshiboun** [comment]
Arun, I tried testing this PR behavior.
Two things I noticed:
on every new run/launch of ros2 node, I'm getting this:
```
[realsense2_camera_node-1] [INFO] [1690716658.785760351] [camera.camera]: Set ROS param depth_module.profile to default: 848x480x30
[realsense2_camera_node-1] [WARN] [1690716658.786094187] [camera.camera]: re-enable the infra stream for the change to take effect.
[realsense2_camera_node-1] [WARN] [1690716658.786197250] [camera.camera]: re-enable the infra1 stream for the change to take effect.
[realsense2_camera_node-1] [WARN] [1690716658.786345386] [camera.camera]: re-enable the infra2 stream for the change to take effect.
[realsense2_camera_node-1] [INFO] [1690716658.794661608] [camera.camera]: Set ROS param rgb_camera.profile to default: 1280x720x30
```

Is this due to this change ?


Also, I left a comment about the parameters names.

File: realsense2_camera/src/profile_manager.cpp [hunk 8]
```diff
@@ -333,12 +406,69 @@ void VideoProfilesManager::registerVideoSensorParams()
...
+                    bool found = false;
+
+                    if (format_str == rs2_format_to_string(RS2_FORMAT_ANY))
+                    {
+                        ROS_WARN_STREAM("A suitable format for " << STREAM_NAME(sip) << " stream will be chosen automatically. "
+                                << "re-enable the stream for the change to take effect.");
+                        found = true;
+                        _formats[sip] = RS2_FORMAT_ANY;
+                    }
+                    else if (string_to_rs2_format(format_str , &temp_format))
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/profile_manager.cpp]: Please change this function name to be something more clear.
I expect it only to convert from string to rs2_format, but it return a Boolean that we need to check.
Or, maybe, change the design of these lines..

File: realsense2_camera/src/ros_utils.cpp [hunk 9]
```diff
@@ -80,6 +80,22 @@ std::string create_graph_resource_name(const std::string &original_name)
     return fixed_name;
 }
 
+bool string_to_rs2_format(std::string str , rs2_format* format)
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/ros_utils.cpp]: I would rather work with 
rs2_format string_to_rs2_format(std::string str)
that returns a RS2_FORMAY_ANY if not found..

**Comment by Arun-Prasad-V** [comment]
>

**Comment by Arun-Prasad-V** [comment]
> Arun, I tried testing this PR behavior. Two things I noticed: on every new run/launch of ros2 node, I'm getting this:
> 
> ```
> [realsense2_camera_node-1] [INFO] [1690716658.785760351] [camera.camera]: Set ROS param depth_module.profile to default: 848x480x30
> [realsense2_camera_node-1] [WARN] [1690716658.786094187] [camera.camera]: re-enable the infra stream for the change to take effect.
> [realsense2_camera_node-1] [WARN] [1690716658.786197250] [camera.camera]: re-enable the infra1 stream for the change to take effect.
> [realsense2_camera_node-1] [WARN] [1690716658.786345386] [camera.camera]: re-enable the infra2 stream for the change to take effect.
> [realsense2_camera_node-1] [INFO] [1690716658.794661608] [camera.camera]: Set ROS param rgb_camera.profile to default: 1280x720x30
> ```
> 
> Is this due to this change ?
> 
> Also, I left a comment about the parameters names.


Hi @SamerKhshiboun ,
This will happen if the default param set by RS ROS2 wrapper is different from the value set during run/launch.

It occurs not only for the new params, also for old params like "rgb_camera.profile" & "depth_module.profile".

File: realsense2_camera/src/profile_manager.cpp (hunk 8 above)
Comment (Arun-Prasad-V) [reply realsense2_camera/src/profile_manager.cpp]: Have updated the design of this function. Will return RS2_FORMAT_ANY if not found. Should I change the function name as well?

**Comment by Arun-Prasad-V** [comment]
> > Arun, I tried testing this PR behavior. Two things I noticed: on every new run/launch of ros2 node, I'm getting this:
> > ```
> > [realsense2_camera_node-1] [INFO] [1690716658.785760351] [camera.camera]: Set ROS param depth_module.profile to default: 848x480x30
> > [realsense2_camera_node-1] [WARN] [1690716658.786094187] [camera.camera]: re-enable the infra stream for the change to take effect.
> > [realsense2_camera_node-1] [WARN] [1690716658.786197250] [camera.camera]: re-enable the infra1 stream for the change to take effect.
> > [realsense2_camera_node-1] [WARN] [1690716658.786345386] [camera.camera]: re-enable the infra2 stream for the change to take effect.
> > [realsense2_camera_node-1] [INFO] [1690716658.794661608] [camera.camera]: Set ROS param rgb_camera.profile to default: 1280x720x30
> > ```
> >     
> >   
> > Is this due to this change ?
> > Also, I left a comment about the parameters names.
> 
> Hi @SamerKhshiboun , This will happen if the default param set by RS ROS2 wrapper is different from the value set during run/launch.
> 
> It occurs not only for the new params, also for old params like "rgb_camera.profile" & "depth_module.profile".

For example - In D455 model:
- For Infra0 stream, there is no default profile. So, ROS2 wrapper set the format to 'UYVY' (first found).
  - But, rs_launch.py sets it to RGB8 format
  - Thus warning message from callback function

**Comment by SamerKhshiboun** [comment]
> > > Arun, I tried testing this PR behavior. Two things I noticed: on every new run/launch of ros2 node, I'm getting this:
> > > ```
> > > [realsense2_camera_node-1] [INFO] [1690716658.785760351] [camera.camera]: Set ROS param depth_module.profile to default: 848x480x30
> > > [realsense2_camera_node-1] [WARN] [1690716658.786094187] [camera.camera]: re-enable the infra stream for the change to take effect.
> > > [realsense2_camera_node-1] [WARN] [1690716658.786197250] [camera.camera]: re-enable the infra1 stream for the change to take effect.
> > > [realsense2_camera_node-1] [WARN] [1690716658.786345386] [camera.camera]: re-enable the infra2 stream for the change to take effect.
> > > [realsense2_camera_node-1] [INFO] [1690716658.794661608] [camera.camera]: Set ROS param rgb_camera.profile to default: 1280x720x30
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Is this due to this change ?
> > > Also, I left a comment about the parameters names.
> > 
> > 
> > Hi @SamerKhshiboun , This will happen if the default param set by RS ROS2 wrapper is different from the value set during run/launch.
> > It occurs not only for the new params, also for old params like "rgb_camera.profile" & "depth_module.profile".
> 
> For example - In D455 model:
> 
> * For Infra0 stream, there is no default profile. So, ROS2 wrapper set the format to 'UYVY' (first found).
>   
>   * But, rs_launch.py sets it to RGB8 format
>   * Thus warning message from callback function

I understand this, so, in this case, the best thing I can think about is, to define 3 parameters for infra, as default formats 
for infra1 and infra2 --> Y8
for infra0 rgb
so, dont take the first found, but takes these if not found.

**Comment by SamerKhshiboun** [comment]
LGTM

**Review by SamerKhshiboun — APPROVED** [review]
